### ISSUE TITLE
Don't output colors when run inside emacs

### ISSUE THREAD

**Issue by thisch** [pre-commit/pre-commit#828]
pre-commit is run by the emacs package magit, which uses a buffer, where TERM=dumb. Unfortunately the current implementation of terminal_supports_color in color.py is True inside emacs. Therefore I see the ANSI color codes of pre-commit in the buffer.

**Comment by asottile** [comment]
I think this is actually a bug with `magit` -- it shouldn't be invoking with a tty if it isn't actually a tty.  I'll look into this though.

**Comment by thisch** [comment]
Thx for the investigation! I'll report this upstream.

**Comment by asottile** [comment]
@thisch yeah I'm banging away in https://github.com/magit/magit/issues/3549 -- some progress there it seems.

It also seems in the meantime you can use the `magit-process-finish-apply-ansi-colors` setting to have magit interpret the escapes itself.  I don't know how to use emacs so when I tried to set this I was pretty lost :laughing:

**Comment by thisch** [comment]
Your were faster ;) Thx. Let's see if this will be fixed in magit.

**Comment by saper** [comment]
This is not as straightforward as it seems. There are utilities that run stuff non-interactively but need to create a tty (`script(1)` for example). Just a simple `isatty()` check is known not to solve the issue permanently. It is especially pity in cases like `script` where one desires to have clean logs without ANSI characters. 

Submitted #1073 to ask about a global ability to disable colors in the output.

**Comment by asottile** [comment]
I'm not sure `script(1)` is a good example -- it does run interactively, and it does capture all terminal symbols (backspaces, line feeds, etc.) and seems to work just fine with colors:

![](https://i.fluffy.cc/HGqmBtKWfghDcpvSSgscVktdbXh60VJX.png)

**Comment by saper** [comment]
Yes, it prints the colors just fine, only the logfile is full of control characters which is not always what one wants.

**Comment by asottile** [comment]
from the man page of script that is intentional:

```
       script places everything in  the  log  file,  including  linefeeds  and
       backspaces.  This is not what the naive user expects.

...

       script  is  primarily designed for interactive terminal sessions. 
```

**Comment by saper** [comment]
fair enough, sorry for noise!

**Comment by asottile** [comment]
np heh :D

**Comment by ahonnecke** [comment]
@thisch 
I was hoping that you might be able to point me to whatever you're using (additional packages, custom config) that makes magit and pre-commit play nicely.

(where "nicely means the failures surface to the magit buffer and I don't have to visit the magit-process buffer)
Right now, failures don't surface to the magit buffer, I have to switch over to the magit-process buffer to see the failures, I'm seeing the same unicode color issues and the file paths are not interactive.

Thanks!

**Comment by thisch** [comment]
> @thisch
> I was hoping that you might be able to point me to whatever you're using (additional packages, custom config) that makes magit and pre-commit play nicely.
> 
> (where "nicely means the failures surface to the magit buffer and I don't have to visit the magit-process buffer)
> Right now, failures don't surface to the magit buffer, I have to switch over to the magit-process buffer to see the failures, I'm seeing the same unicode color issues and the file paths are not interactive.
> 
> Thanks!

I have to disappoint you, since I don't use pre-commit anymore in my emacs setup.
